### PR TITLE
Newbuild

### PR DESCRIPTION
**Pull request by GordonSmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dockerfiles/vcpkg/build.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    docker build --progress plain --pull --rm -f "$SCRIPT_DIR/$1.dockerfile" \
+        -t build-$1:$GITHUB_REF \
+        -t build-$1:latest \
+        --build-arg VCPKG_REF=$VCPKG_REF \
+        "$SCRIPT_DIR/." 
+
+    docker run --rm --mount source="$(pwd)",target=/hpcc-dev/HPCC-Platform,type=bind,consistency=cached build-$1:$GITHUB_REF \
+        "cmake -S /hpcc-dev/HPCC-Platform -B /hpcc-dev/HPCC-Platform/build-$1 \${CMAKE_OPTIONS}"
+
+    docker run --rm --mount source="$(pwd)",target=/hpcc-dev/HPCC-Platform,type=bind,consistency=cached build-$1:$GITHUB_REF \
```
Comment (jakesmith) [inline dockerfiles/vcpkg/build.sh:37]: minor: as per previous comments, could reuse same container afaics.

**Review by jakesmith — APPROVED** [review]
@GordonSmith - see comments, but nothing to stop approving

**Review by ghalliday — APPROVED** [review]

File: .github/workflows/build-assets.yml [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Create Build Image
+        uses: docker/build-push-action@v3
+        with:
+          builder: ${{ steps.buildx.outputs.name }}
+          file: dockerfiles/vcpkg/${{ matrix.os }}.dockerfile
+          context: dockerfiles/vcpkg
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:94]: I agree - if only for consistency with existing ones (was on my mental todo list):  https://track.hpccsystems.com/browse/HPCC-28797

File: .github/workflows/build-assets.yml [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+          push: true
+          tags: |
+            ${{ secrets.DOCKER_USERNAME }}/build-${{ matrix.os }}:${{ steps.vars.outputs.branch_name }} 
+            ${{ secrets.DOCKER_USERNAME }}/build-${{ matrix.os }}:latest
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:98]: Agree - also there is probably no point on even pushing these to dockerhub until we have a (re)use for them:  https://track.hpccsystems.com/browse/HPCC-28798

File: .github/workflows/build-assets.yml [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+          sudo rm -f ./build/CMakeCache.txt
+          sudo rm -rf ./build/CMakeFiles
+          docker_label=${{ secrets.DOCKER_USERNAME }}/build-${{ matrix.os }}:${{ steps.vars.outputs.branch_name }}
+          docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake -S /hpcc-dev/HPCC-Platform -B /hpcc-dev/HPCC-Platform/build ${{ steps.vars.outputs.cmake_options }} -DINCLUDE_PLUGINS=ON -DCONTAINERIZED=ON -DSUPPRESS_REMBED=ON -DSUPPRESS_V8EMBED=ON -DSUPPRESS_SPARK=ON -DCPACK_STRIP_FILES=OFF"
+          docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake --build /hpcc-dev/HPCC-Platform/build --parallel $(nproc) --target package" 
+          docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake -S /hpcc-dev/HPCC-Platform -B /hpcc-dev/HPCC-Platform/build ${{ steps.vars.outputs.cmake_options }} -DINCLUDE_PLUGINS=ON -DCONTAINERIZED=ON -DSUPPRESS_REMBED=ON -DSUPPRESS_V8EMBED=ON -DSUPPRESS_SPARK=ON -DCPACK_STRIP_FILES=ON" 
+          docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake --build /hpcc-dev/HPCC-Platform/build --parallel $(nproc) --target package"
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:129]: Could do - might make it harder to read here though?

File: .github/workflows/build-assets.yml [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+          token: ${{ secrets.LNB_TOKEN }}
+          allowUpdates: true
+          artifacts: "build/hpccsystems-platform-internal*.deb,build/hpccsystems-platform-internal*.rpm"
+
+      - name: Upload error logs
+        if: ${{ failure() || cancelled() }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: ${{ matrix.os }}-${{ matrix.package }}-logs
+          path: build/CMakeCache.txt
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:173]: Good catch I had switched from a logs grep to this while testing something weird:  https://track.hpccsystems.com/browse/HPCC-28799

File: dockerfiles/vcpkg/build.sh (hunk 1 above)
Comment (GordonSmith) [reply dockerfiles/vcpkg/build.sh:37]: Again I would need to see what that looks like...

File: .github/workflows/build-assets.yml [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            VCPKG_REF=${{ steps.vars.outputs.vcpkg_sha_short }}
+            BUILD_FOLDER=build-${{ matrix.os }}
+
+      - name: CMake Packages
+        if: ${{ !matrix.container && !matrix.ln }}
+        run: |
+          plugins=("CASSANDRAEMBED" "COUCHBASEEMBED" "ECLBLAS" "H3" "JAVAEMBED" "KAFKA" "MEMCACHED" "MYSQLEMBED" "NLP" "REDIS" "SQLITE3EMBED" "SQS" "PLATFORM")
+          for plugin in "${plugins[@]}"; do
+            sudo rm -f ./build/CMakeCache.txt
+            sudo rm -rf ./build/CMakeFiles
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:112]: No this has nothing to do with ownership of the build folder.
Each iteration its setting 1 plugin to "ON", this has a cumulative effect on the CMakeCache.txt file (IOW appending an additional plugin each iteration).  Clearing the cache fixes this, while keeping the previously built libs.

File: .github/workflows/build-assets.yml [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+            docker_label=${{ secrets.DOCKER_USERNAME }}/build-${{ matrix.os }}:${{ steps.vars.outputs.branch_name }}
+            docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake -S /hpcc-dev/HPCC-Platform -B /hpcc-dev/HPCC-Platform/build ${{ steps.vars.outputs.cmake_options }} -D$plugin=ON -DCONTAINERIZED=OFF -DCPACK_STRIP_FILES=OFF" 
+            docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake --build /hpcc-dev/HPCC-Platform/build --parallel $(nproc) --target package" 
+            docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake -S /hpcc-dev/HPCC-Platform -B /hpcc-dev/HPCC-Platform/build ${{ steps.vars.outputs.cmake_options }} -D$plugin=ON -DCONTAINERIZED=OFF -DCPACK_STRIP_FILES=ON"
+            docker run --rm --mount ${{ steps.vars.outputs.mount_platform }} $docker_label "cmake --build /hpcc-dev/HPCC-Platform/build --parallel $(nproc) --target package"
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply .github/workflows/build-assets.yml:117]: Would want to see what that looks like from a readability point of view

**Pull request merged**